x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/tool/BaseClientAwareCommand.java

```diff
@@ -0,0 +1,150 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.security.tool;
+
+import org.elasticsearch.Version;
+import org.elasticsearch.cli.ExitCodes;
+import org.elasticsearch.cli.KeyStoreAwareCommand;
+import org.elasticsearch.cli.Terminal;
+import org.elasticsearch.cli.UserException;
+import org.elasticsearch.common.io.Streams;
+import org.elasticsearch.common.settings.KeyStoreWrapper;
+import org.elasticsearch.common.settings.SecureString;
+import org.elasticsearch.core.CheckedFunction;
+import org.elasticsearch.env.Environment;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.MalformedURLException;
+import java.net.URISyntaxException;
+import java.net.URL;
+import java.security.SecureRandom;
+import java.util.Objects;
+import java.util.function.Function;
+
+/**
+ * A {@link KeyStoreAwareCommand} that can be extended for any CLI tool that needs use a http client.
+ */
+public abstract class BaseClientAwareCommand extends KeyStoreAwareCommand {
+    protected final Function<Environment, CommandLineHttpClient> clientFunction;
+    protected final CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction;
+    final SecureRandom secureRandom = new SecureRandom();
+
+    public BaseClientAwareCommand(
+        Function<Environment, CommandLineHttpClient> clientFunction,
+        CheckedFunction<Environment, KeyStoreWrapper, Exception> keyStoreFunction,
+        String description
+    ) {
+        super(description);
+        this.clientFunction = clientFunction;
+        this.keyStoreFunction = keyStoreFunction;
+    }
+
+    /**
+     * Checks that we can connect to the cluster and that the cluster health is not RED. It optionally handles
+     * retries as the file realm might not have reloaded the users file yet in order to authenticate our
+     * newly created file realm user.
+     */
+    protected void checkClusterHealthWithRetries(CommandLineHttpClient client, Terminal terminal, String username, SecureString password,
+                                                 int retriesWaitingForUser, int retriesWaitingForNode, boolean force) throws Exception {
```

Comment:
Two types of retries introduced here. They can be used together or separately, depending on the caller logic. Waiting for the user will only retry if authentication fails. Waiting for the node will retry if connection is refused or node is Red (making N retries all together, for ex. try to time until connection is established and 3 more if the cluster is Red after it)